aggregator/src/aggregation/rlc/gates.rs

```diff
@@ -0,0 +1,427 @@
+        for index in *offset..(1 << LOG_DEGREE) - 1 {
+            region.assign_advice(
+                || "pad",
+                self.phase_2_column,
+                index,
+                || Value::known(Fr::zero()),
+            )?;
+        }
+        Ok(())
+    }
+
+    // decompose a field element into 254 bits of boolean cells
+    pub(crate) fn decomposition(
+        &self,
+        region: &mut Region<Fr>,
+        input: &AssignedCell<Fr, Fr>,
+        offset: &mut usize,
+    ) -> Result<Vec<AssignedCell<Fr, Fr>>, Error> {
+        let mut input_element = Fr::default();
+        input.value().map(|&x| input_element = x);
+
+        let bits = input_element
+            .to_bytes()
+            .iter()
+            .flat_map(byte_to_bits_le)
+            .collect::<Vec<_>>();
+        // sanity check
+        {
+            let mut reconstructed = Fr::zero();
+            bits.iter().rev().for_each(|bit| {
+                reconstructed *= Fr::from(2);
+                reconstructed += Fr::from(*bit as u64);
+            });
+            assert_eq!(reconstructed, input_element);
+        }
+
+        let bit_cells = bits
+            .iter()
+            .take(254) // hard coded for BN curve
+            .map(|&bit| {
+                let cell = self.load_private(region, &Fr::from(bit as u64), offset)?;
+                self.enforce_binary(region, &cell, offset)?;
+                Ok(cell)
+            })
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        let mut acc = {
+            let zero = self.load_private(region, &Fr::from(0), offset)?;
+            let zero_cell = self.zero_cell(zero.cell().region_index);
+            region.constrain_equal(zero_cell, zero.cell())?;
+            zero
+        };
+
+        let two = {
+            let two = self.load_private(region, &Fr::from(2), offset)?;
+            let two_cell = self.two_cell(two.cell().region_index);
+            region.constrain_equal(two_cell, two.cell())?;
+            two
+        };
+
+        for bit in bit_cells.iter().rev() {
+            acc = self.mul_add(region, &acc, &two, bit, offset)?;
+        }
+
+        // sanity check
+        assert_equal(&acc, input);
+
+        region.constrain_equal(acc.cell(), input.cell())?;
+
+        Ok(bit_cells)
+    }
+
+    // return a boolean if a is smaller than b
+    // requires that both a and b are smallish
+    pub(crate) fn is_smaller_than(
+        &self,
+        region: &mut Region<Fr>,
+        a: &AssignedCell<Fr, Fr>,
+        b: &AssignedCell<Fr, Fr>,
+        offset: &mut usize,
+    ) -> Result<AssignedCell<Fr, Fr>, Error> {
+        // when a and b are both small (as in our use case)
+        // if a < b, (a-b) will under flow and the highest bit of (a-b) be one
+        // else,  the highest bit of (a-b) be zero
+        let sub = self.sub(region, a, b, offset)?;
+        let bits = self.decomposition(region, &sub, offset)?;
+        Ok(bits[253].clone())
+    }
+}
```

Comment:
This function is **buggy** because it uses `decomposition`. Any integer can have two different decompositions of 254 bits. For example, 
1. `0 = LC_BE([0, ..., 0])`;
2. `0 = LC_BE([1, 0, 0, 1, 0, 1, ...., 1, 1, 0, 1])`.

Therefore an malicious attacker can convince the verifier that the highest bit of a - b is 1 when a >= b.

Reply:
I suggest that we add one more argument `nbits` in this function to signify that  both operands are within `[0, 2^nbits)` and then we can decompose `2^nbits + a - b` into `(nbits + 1)` bits. The highest bit will be returned as the result of `is_smaller_than` function.

If a < b, then the highest bit must be 0; else, the highest bit must be true.